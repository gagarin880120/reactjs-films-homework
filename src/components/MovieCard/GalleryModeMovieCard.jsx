import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import styles from './MovieCard.module.scss';
import InfoButton from '../InfoButton/InfoButton';
import TrailerButtonRect from '../TrailerButtonRect/TrailerButtonRect';
import TrailerButtonRound from '../TrailerButtonRound/TrailerButtonRound';

export default function GalleryModeMovieCard({
  movie, onTrailerButtonClick, onLinkClick,
}) {
  const [isInfoViewed, setIsInfoViewed] = useState(false);
  const hoverEl = useRef();
  const infoEl = useRef();

  return (
    <div
      className={styles.galleryWrapper}
      style={{
        backgroundImage: movie.poster_path
          ? `url(https://image.tmdb.org/t/p/w300/${movie.poster_path})`
          : null,
      }}
    >
      <div
        ref={hoverEl}
        className={styles.hoverTrailer}
      >
        <TrailerButtonRound
          onTrailerButtonClick={onTrailerButtonClick}
          id={movie.id}
        />
        <span className={styles.hoverTrailerText}>Watch Now</span>
        <InfoButton
          onInfoButtonClick={() => {
            hoverEl.current.className = styles.hoverElDisplayNone;
            setIsInfoViewed(true);
            infoEl.current.className = styles.infoIsViewed;
          }}
        />
      </div>
      <div
        className={styles.movieInfo}
        ref={infoEl}
      >
        {
          isInfoViewed ? (
            <button
              className={styles.closeInfo}
              type="button"
              onClick={() => {
                setIsInfoViewed(false);
                infoEl.current.className = styles.movieInfo;
                hoverEl.current.className = styles.hoverTrailer;
              }}
            >
              <i className="fas fa-times" />
            </button>
          ) : null
        }
        <Link
          to={`/movie/${movie.id}`}
          onClick={() => {
            onLinkClick(movie.id);
          }}
        >
          <p className={styles.movieTitle}>{movie.title}</p>
        </Link>
        <span className={styles.movieRating}>{movie.vote_average}</span>
        <p className={styles.movieGenre}>{movie.genres}</p>
        {
          isInfoViewed
            ? (
              <>
                <div className={styles.overview}>{movie.overview}</div>
                <TrailerButtonRect
                  onTrailerButtonClick={onTrailerButtonClick}
                  id={movie.id}
                />
              </>
            ) : null
        }
      </div>
    </div>
  );
}

GalleryModeMovieCard.propTypes = {
  onTrailerButtonClick: PropTypes.func,
  onLinkClick: PropTypes.func,
  movie: PropTypes.shape({
    poster_path: PropTypes.string,
    id: PropTypes.number,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    genres: PropTypes.string,
    overview: PropTypes.string,
  }),
};

GalleryModeMovieCard.defaultProps = {
  onTrailerButtonClick: null,
  onLinkClick: null,
  movie: {
    poster_path: '',
    id: 0,
    title: '',
    vote_average: 0,
    genres: '',
    overview: '',
  },
};
