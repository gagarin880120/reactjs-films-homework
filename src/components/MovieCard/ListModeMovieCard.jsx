import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import styles from './MovieCard.module.scss';
import TrailerButtonRect from '../TrailerButtonRect/TrailerButtonRect';

export default function ListModeMovieCard({
  movie, onTrailerButtonClick, onLinkClick,
}) {
  return (
    <div
      className={styles.listWrapper}
    >
      <div
        className={styles.listImage}
        style={{
          backgroundImage: movie.poster_path
            ? `url(https://image.tmdb.org/t/p/w300/${movie.poster_path})`
            : null,
        }}
      />
      <div className={styles.listMovieInfo}>
        <Link
          to={`/movie/${movie.id}`}
          onClick={() => {
            onLinkClick(movie.id);
          }}
        >
          <p className={styles.movieTitle}>{movie.title}</p>
        </Link>
        <span className={styles.movieRating}>{movie.vote_average}</span>
        <p className={styles.movieGenre}>{movie.genres}</p>

        <div className={styles.overview}>{movie.overview}</div>
        <TrailerButtonRect
          onTrailerButtonClick={onTrailerButtonClick}
          id={movie.id}
        />
      </div>
    </div>
  );
}

ListModeMovieCard.propTypes = {
  onTrailerButtonClick: PropTypes.func,
  onLinkClick: PropTypes.func,
  movie: PropTypes.shape({
    poster_path: PropTypes.string,
    id: PropTypes.number,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    genres: PropTypes.string,
    overview: PropTypes.string,
  }),
};

ListModeMovieCard.defaultProps = {
  onTrailerButtonClick: null,
  onLinkClick: null,
  movie: {
    poster_path: '',
    id: 0,
    title: '',
    vote_average: 0,
    genres: '',
    overview: '',
  },
};
