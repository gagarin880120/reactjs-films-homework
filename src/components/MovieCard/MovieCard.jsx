import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import styles from './MovieCard.module.scss';
import InfoButton from '../InfoButton/InfoButton';
import TrailerButtonRect from '../TrailerButtonRect/TrailerButtonRect';
import TrailerButtonRound from '../TrailerButtonRound/TrailerButtonRound';

export default function MovieCard({
  movie, onTrailerButtonClick, onLinkClick,
}) {
  const [isInfoViewed, setIsInfoViewed] = useState(false);
  const hoverEl = useRef();
  const infoEl = useRef();

  return (
    <div
      className={styles.wrapper}
      style={{
        backgroundImage: movie.poster_path
          ? `url(https://image.tmdb.org/t/p/w185/${movie.poster_path})`
          : null,
      }}
    >
      <div
        ref={hoverEl}
        className={styles.hoverTrailer}
      >
        <TrailerButtonRound
          onTrailerButtonClick={onTrailerButtonClick}
          id={movie.id}
        />
        <span className={styles.hoverTrailerText}>Watch Now</span>
        <InfoButton
          onInfoButtonClick={() => {
            hoverEl.current.className = styles.hoverElDisplayNone;
            setIsInfoViewed(true);
            infoEl.current.className = styles.infoIsViewed;
          }}
        />
      </div>
      <div
        className={styles.movieInfo}
        ref={infoEl}
      >
        {
          isInfoViewed ? (
            <button
              className={styles.closeInfo}
              type="button"
              onClick={() => {
                setIsInfoViewed(false);
                infoEl.current.className = styles.movieInfo;
                hoverEl.current.className = styles.hoverTrailer;
              }}
            >
              X
            </button>
          ) : null
        }
        <Link
          to="/movieDetailsPage"
          onClick={() => {
            onLinkClick(movie.id);
          }}
        >
          <p className={styles.movieTitle}>{movie.title}</p>
        </Link>
        <span className={styles.movieRating}>{movie.vote_average}</span>
        <p className={styles.movieGenre}>{movie.genres}</p>
        {
          isInfoViewed
            ? (
              <>
                <div className={styles.overview}>{movie.overview}</div>
                <TrailerButtonRect
                  onTrailerButtonClick={onTrailerButtonClick}
                  id={movie.id}
                />
              </>
            ) : null
        }
      </div>
    </div>
  );
}

MovieCard.propTypes = {
  onTrailerButtonClick: PropTypes.func,
  onLinkClick: PropTypes.func,
  movie: PropTypes.shape({
    poster_path: PropTypes.string,
    id: PropTypes.number,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    genres: PropTypes.string,
    overview: PropTypes.string,
  }),
};

MovieCard.defaultProps = {
  onTrailerButtonClick: null,
  onLinkClick: null,
  movie: {
    poster_path: '',
    id: 0,
    title: '',
    vote_average: 0,
    genres: '',
    overview: '',
  },
};
