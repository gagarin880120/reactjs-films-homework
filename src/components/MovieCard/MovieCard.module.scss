.wrapper {
  background-size: cover;
  width: 280px;
  height: 340px;
  position: relative;
}

.wrapper:hover .hoverTrailer {
  display: flex;
}

.hoverElDisplayNone {
  display: none;
}

.movieInfo {
  height: 70px;
  margin-top: 240px;
  background: #fff;
  padding: 30px 5% 0;
}

.movieTitle {
  white-space: nowrap;
  overflow: hidden;
  width: 210px;
  color: #242b2d;
  text-transform: uppercase;
  font-weight: bold;
  text-overflow: ellipsis;
}

.movieRating {
  float: right;
  color: #47c3f0;
  border: 1px solid #47c3f0;
  border-radius: 4px;
  padding: 2px 5px;
  margin-top: -20px;
}

.movieGenre {
  margin-top: 10px;
  font-size: 15px;
  color: #47c3f0;
  font-weight: bold;
}

.hoverTrailer {
  width: 280px;
  height: 180px;
  background: rgba(0, 0, 0, 0.6);
  display: none;
  position: absolute;
  top: 0;
  left: 0;
  flex-direction: column;
  align-items: center;
  padding-top: 60px;
}

.hoverTrailerText {
  color: #fff;
  margin: 10px 0 20px;
}

.infoIsViewed {
  background: rgba(0, 0, 0, 0.7);
  height: 300px;
  padding: 30px 7% 10px;
  overflow-y: scroll;

  .movieTitle {
    color: #fff;
  }
}

.infoIsViewed::-webkit-scrollbar {
  width: 8px;
}

.infoIsViewed::-webkit-scrollbar-track {
  background: rgba(0, 0, 0, 0.5);
  border-radius: 6px;
}

.infoIsViewed::-webkit-scrollbar-thumb {
  background: #47c3f0;
  border-radius: 6px;
}

.infoIsViewed::-webkit-scrollbar-thumb:hover {
  background: #02729b;
}

.closeInfo {
  padding-bottom: 20px;

  span {
    cursor: pointer;
  }
}

.overview {
  margin: 20px 0;
  font-size: 15px;
  line-height: 20px;
}
