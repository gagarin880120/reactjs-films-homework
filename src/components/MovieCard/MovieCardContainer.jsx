import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieCard from './MovieCard';
import { setModal, getTrailer } from '../../redux/actions';

export function MovieCardContainer({ onTrailerButtonClick, movie }) {
  return (
    <MovieCard
      onTrailerButtonClick={onTrailerButtonClick}
      movie={movie}
    />
  );
}

export const mapDispatchToProps = (dispatch) => ({
  onTrailerButtonClick(id) {
    dispatch(setModal(true));
    dispatch(getTrailer(id));
  },
});

MovieCardContainer.propTypes = {
  onTrailerButtonClick: PropTypes.func,
  movie: PropTypes.shape({}),
};

MovieCardContainer.defaultProps = {
  onTrailerButtonClick: null,
  movie: {},
};

export default connect(null, mapDispatchToProps)(MovieCardContainer);
