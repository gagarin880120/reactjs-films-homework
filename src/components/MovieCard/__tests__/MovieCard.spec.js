import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer, { act } from 'react-test-renderer';
import React from 'react';
import MovieCard from '../MovieCard';

describe('MovieCard component', () => {
  const movie = {
    title: 'Star Wars',
    genre_ids: [28],
    id: 11
  }

  const movieWithPoster = {
    title: 'Star Wars',
    genre_ids: [28],
    id: 11,
    poster_path: 'path'
  }

  const genres = [{id:28, name:'Adventure'}]

  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MovieCard movie={movie} genres={genres} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  describe('with mocked refs', () => {
    const testRenderer = TestRenderer.create(
      <MovieCard movie={movieWithPoster} genres={genres} />,
      {
        createNodeMock: (element) => {
          return element;
        }
      }
    );

    test('when infoButton is clicked should render overview element', () => {
      const button = testRenderer.root.findByProps({className: 'infoButton'});

      act(() => {
        button.props.onClick();
      });

      const overview = testRenderer.root.findByProps({className: 'overview'});

      expect(overview).not.toBeNull();
    });

    test('when closeInfo is clicked should change hoverEl className to hoverTrailer', () => {
      const closeInfo = testRenderer.root.findByProps({className: 'closeInfo'});
      const hoverEl = testRenderer.root.findByProps({className: 'hoverTrailer'});

      act(() => {
        closeInfo.props.onClick();
      });

      expect(hoverEl.props.className).toBe('hoverTrailer');
    });
  });
});
