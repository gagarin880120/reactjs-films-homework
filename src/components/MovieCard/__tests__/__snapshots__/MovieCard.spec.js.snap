// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component should render without crashing 1`] = `
<React.Fragment>
  <div
    className="wrapper"
    style={
      Object {
        "backgroundImage": "url(https://spnbag.com/assets/general/images/no_poster.jpg)",
      }
    }
  >
    <div
      className="hoverTrailer"
      testid="hoverTrailer"
    >
      <TrailerButtonRound
        id={11}
        onTrailerButtonClick={null}
      />
      <span
        className="hoverTrailerText"
      >
        Watch Now
      </span>
      <InfoButton
        onInfoButtonClick={[Function]}
      />
    </div>
    <div
      className="movieInfo"
      testid="info"
    >
      <p
        className="movieTitle"
      >
        Star Wars
      </p>
      <span
        className="movieRating"
      />
      <p
        className="movieGenre"
      />
    </div>
  </div>
</React.Fragment>
`;
